golem-common/src/virtual_exports/http_incoming_handler.rs

```diff
@@ -0,0 +1,524 @@
+            Self::DELETE => Ok(http::Method::DELETE),
+            Self::CONNECT => Ok(http::Method::CONNECT),
+            Self::OPTIONS => Ok(http::Method::OPTIONS),
+            Self::TRACE => Ok(http::Method::TRACE),
+            Self::PATCH => Ok(http::Method::PATCH),
+            Self::Custom(method) => http::Method::from_bytes(method.as_bytes()),
+        }
+    }
+}
+
+pub struct HttpFields(pub Vec<(String, Bytes)>);
+
+impl HttpFields {
+    pub fn analyzed_type() -> AnalysedType {
+        use golem_wasm_ast::analysis::*;
+        AnalysedType::List(TypeList {
+            inner: Box::new(AnalysedType::Tuple(TypeTuple {
+                items: vec![
+                    AnalysedType::Str(TypeStr),
+                    AnalysedType::List(TypeList {
+                        inner: Box::new(AnalysedType::U8(TypeU8)),
+                    }),
+                ],
+            })),
+        })
+    }
+
+    pub fn from_value(value: &Value) -> Result<Self, String> {
+        let mut result = Vec::new();
+
+        let list_values = extract!(value, Value::List(inner), inner, "not a list")?;
+
+        for lv in list_values {
+            let tuple_value = extract!(lv, Value::Tuple(inner), inner, "not a tuple")?;
+
+            let (name, values) = extract!(
+                tuple_value.as_slice(),
+                [Value::String(name), Value::List(values)],
+                (name.clone(), values),
+                "incompatible types"
+            )?;
+
+            let mut result_value = Vec::new();
+
+            for v in values {
+                let v = extract!(v, Value::U8(inner), *inner, "not a byte")?;
+                result_value.push(v);
+            }
+
+            result.push((name, Bytes::from(result_value)));
+        }
+
+        Ok(HttpFields(result))
+    }
+
+    pub fn to_value(self) -> Value {
+        let mut list_values = Vec::new();
+
+        for (name, value) in self.0 {
+            let converted_bytes: Vec<Value> = value.into_iter().map(Value::U8).collect::<Vec<_>>();
+
+            list_values.push(Value::Tuple(vec![
+                Value::String(name),
+                Value::List(converted_bytes),
+            ]));
+        }
+        Value::List(list_values)
+    }
+}
+
+pub struct HttpBodyContent(pub Bytes);
+
+impl HttpBodyContent {
+    pub fn analyzed_type() -> AnalysedType {
+        use golem_wasm_ast::analysis::*;
+        AnalysedType::List(TypeList {
+            inner: Box::new(AnalysedType::U8(TypeU8)),
+        })
+    }
+
+    pub fn from_value(value: &Value) -> Result<Self, String> {
+        let mut result = Vec::new();
+
+        let list_values = extract!(value, Value::List(inner), inner, "not a list")?;
+
+        for lv in list_values {
+            let byte_value = extract!(lv, Value::U8(inner), *inner, "not a byte")?;
+            result.push(byte_value);
+        }
+
+        Ok(HttpBodyContent(Bytes::from(result)))
+    }
+
+    pub fn to_value(self) -> Value {
```

Comment:
Instead of these `to_value` calls consider implementing the `IntoValue` type class (from wasm-rpc), that also requires you to specify the type, but it will allow us to use these in more scenarios (for example better integrating it `PublicOplogEntry` etc)

Reply:
Annoyance here is that only one half of the types here need to go to values (the other half is being converted from values), while we need to have the type for all of them.

This would suggest that we need a counterpart of IntoValue that can go from values (FromValue?) and then we would likely want to split the type portions out from both.

I wanted to avoid doing this for now and just implemented the needed functionality as plain functions. We can always implement IntoValue if we end up needing it for a specific type.